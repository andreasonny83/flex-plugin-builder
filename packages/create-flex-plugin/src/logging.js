import chalk from 'chalk';
import wrapAnsi from 'wrap-ansi';
import * as symbols from 'log-symbols';
import { stripIndent } from 'common-tags';
import boxen from 'boxen';

const ERROR = chalk.bold('ERROR');
const headline = chalk.bold.green;

export function error(message, ...args) {
  console.error(`${symbols.error} ${ERROR} ${message}`, ...args);
}

export function finalMessage(config) {
  const tool = config.yarn ? 'yarn' : 'npm';

  const installCommand = config.yarn ? 'yarn' : 'npm install';
  const setupMessage = stripIndent`
    ${headline('Setup:')}
      $ cd ${config.pluginFileName}/
      $ ${installCommand}
  `;

  const startCommand = `${tool} start`;
  const devMessage = stripIndent`
    ${headline('Development:')}
      $ cd ${config.pluginFileName}/
      $ ${startCommand}
  `;

  const buildCommand = config.yarn ? 'yarn build' : 'npm run build';
  const buildMessage = stripIndent`
    ${headline('Build Command:')}
      $ cd ${config.pluginFileName}/
      $ ${buildCommand}
  `;

  let message = stripIndent`
    Your Twilio Flex plugin project has been successfully created!${
      !config.install ? `\n\n  ${setupMessage}` : ''
    }

  ${devMessage}

  ${buildMessage}

  For more info check the README.md file or go to:
  âž¡ https://www.twilio.com/docs/flex
  `;

  const columns = (process.stdout.columns || 100) - 14 /* boxen spacing */;
  message = wrapAnsi(message, columns, { hard: true });

  const boxedMessage = boxen(message, {
    padding: 1,
    margin: 1,
  });
  console.log(boxedMessage);
}
