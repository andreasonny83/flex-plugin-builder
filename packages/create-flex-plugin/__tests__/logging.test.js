import * as log from '../src/logging';
import * as symbols from 'log-symbols';
import chalk from 'chalk';

describe('logging', () => {
  // Mocking the console object
  global.console = {
    log: jest.fn(),
    error: jest.fn(),
  }

  beforeEach(() => jest.clearAllMocks());

  describe('error', () => {
    const SYMBOL_ERROR = symbols.error;
    const ERROR = chalk.bold('ERROR');

    test(`should return a 'console.error' message`, () => {
      // Arrange
      const errorMessage = 'Error message';
      const expectedMessage = `${SYMBOL_ERROR} ${ERROR} ${errorMessage}`;

      // Act
      log.error(errorMessage);

      // Assert
      expect(console.error).toHaveBeenCalledWith(expectedMessage);
    });

    test(`should accept any number of arguments`, () => {
      // Arrange
      const errorMessage = 'Error message';
      const arg1 = 'argument #1';
      const arg2 = 'argument #2';
      const args = [arg1, arg2];
      const expectedMessage = `${SYMBOL_ERROR} ${ERROR} ${errorMessage}`;

      // Act
      log.error(errorMessage, ...args);

      // Assert
      expect(console.error).toHaveBeenCalledWith(expectedMessage, ...args);
    });
  });

  describe('finalMessage', () => {
    test(`should render an 'npm' setup message to the console`, () => {
      // Arrange
      let message;
      console.log = jest.fn((msg) => message = msg);

      // Act
      log.finalMessage({});

      // Assert
      expect(console.log).toHaveBeenCalled();
      expect(message).toMatchSnapshot();
    });

    test(`should render a 'yarn' setup message to the console`, () => {
      // Arrange
      let message;
      console.log = jest.fn((msg) => message = msg);

      // Act
      log.finalMessage({yarn: true});

      // Assert
      expect(console.log).toHaveBeenCalled();
      expect(message).toMatchSnapshot();
    });

    test(`should render an instruction message skipping the setup step`, () => {
      // Arrange
      let message;
      console.log = jest.fn((msg) => message = msg);

      // Act
      log.finalMessage({install: true});

      // Assert
      expect(console.log).toHaveBeenCalled();
      expect(message).toMatchSnapshot();
    });
  });
});