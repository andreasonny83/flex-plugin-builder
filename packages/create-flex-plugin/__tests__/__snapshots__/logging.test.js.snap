// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logging finalMessage should render a 'yarn' setup message to the console 1`] = `
"Your Twilio Flex plugin project has been successfully created!

[1m[32mSetup:[39m[22m
$ cd undefined/
$ yarn

[1m[32mDevelopment:[39m[22m
$ cd undefined/
$ yarn start

[1m[32mBuild Command:[39m[22m
$ cd undefined/
$ yarn build

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;

exports[`logging finalMessage should render an 'npm' setup message to the console 1`] = `
"Your Twilio Flex plugin project has been successfully created!

[1m[32mSetup:[39m[22m
$ cd undefined/
$ npm install

[1m[32mDevelopment:[39m[22m
$ cd undefined/
$ npm start

[1m[32mBuild Command:[39m[22m
$ cd undefined/
$ npm run build

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;

exports[`logging finalMessage should render an instruction message skipping the setup step 1`] = `
"Your Twilio Flex plugin project has been successfully created!

[1m[32mDevelopment:[39m[22m
$ cd undefined/
$ npm start

[1m[32mBuild Command:[39m[22m
$ cd undefined/
$ npm run build

For more info check the README.md file or go to:
âž¡ https://www.twilio.com/docs/flex"
`;
